sentry/src/main/java/io/sentry/instrumentation/file/SentryFileInputStream.java

```diff
@@ -0,0 +1,197 @@
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * An implementation of {@link java.io.FileInputStream} that creates a {@link io.sentry.ISpan} for
+ * reading operation with filename and byte count set as description
+ *
+ * Note, that span is started when this InputStream is instantiated via constructor and finishes
+ * when the {@link java.io.FileInputStream#close()} is called.
+ */
+@SuppressWarnings("unused") // used via bytecode manipulation (SAGP)
+@Open
+public class SentryFileInputStream extends FileInputStream {
+
+  private final @Nullable ISpan currentSpan;
+  private final @Nullable File file;
+  private final @NotNull FileInputStream delegate;
+
+  private @NotNull SpanStatus spanStatus = SpanStatus.OK;
+  private long byteCount;
+
+  public SentryFileInputStream(@Nullable String name) throws FileNotFoundException {
+    this(init(name != null ? new File(name) : null, null));
+  }
+
+  public SentryFileInputStream(@Nullable File file) throws FileNotFoundException {
+    this(init(file, null));
+  }
+
+  public SentryFileInputStream(FileDescriptor fdObj) {
+    this(init(fdObj, null), fdObj);
+  }
+
+  private SentryFileInputStream(
+    @NotNull FileInputStreamInitData data,
+    @NotNull FileDescriptor fd
+  ) {
+    super(fd);
+    file = null;
+    currentSpan = data.span;
+    delegate = data.delegate;
+  }
+
+  private SentryFileInputStream(
+    @NotNull FileInputStreamInitData data
+  ) throws FileNotFoundException {
+    super(data.file);
+    currentSpan = data.span;
+    file = data.file;
+    delegate = data.delegate;
+  }
+
+  private static FileInputStreamInitData init(
+    @Nullable File file,
+    @Nullable FileInputStream delegate
+  ) throws FileNotFoundException {
+    final ISpan span = startSpan();
+    if (delegate == null) {
+      delegate = new FileInputStream(file);
+    }
+    return new FileInputStreamInitData(file, span, delegate);
+  }
+
+  private static FileInputStreamInitData init(
+    @NotNull FileDescriptor fd,
+    @Nullable FileInputStream delegate
+  ) {
+    final ISpan span = startSpan();
+    if (delegate == null) {
+      delegate = new FileInputStream(fd);
+    }
+    // TODO: it's only possible to get filename from FileDescriptor via reflection AND when it's
+    // TODO: running on Android, should we do that?
+    return new FileInputStreamInitData(null, span, delegate);
+  }
+
+  private static @Nullable
+  ISpan startSpan() {
+    final ISpan parent = Sentry.getSpan();
+    return parent != null ? parent.startChild("file.read") : null;
+  }
+
+  @Override
+  public int read() throws IOException {
+    try {
+      int result = delegate.read();
+      if (result != -1) {
+        byteCount++;
+      }
+      return result;
+    } catch (IOException exception) {
+      spanStatus = SpanStatus.INTERNAL_ERROR;
+      throw exception;
+    }
+  }
+
+  @Override
+  public int read(byte @NotNull [] b) throws IOException {
```

Comment:
```suggestion
  public int read(@NotNull byte [] b) throws IOException {
```

nit only, annotations always before

Reply:
I think it has to be where it is now, otherwise it has no effect
![image](https://user-images.githubusercontent.com/4999776/144588690-8ae8c358-eb23-4271-af85-e1d49b0f00ea.png)